src/main/java/org/truffleruby/builtins/AmbiguousOptionalArgumentChecker.java

```diff
@@ -36,54 +40,83 @@ public static void verifyNoAmbiguousOptionalArguments(NodeFactory<? extends Ruby
-                                unguardedObjectArgument = true;
-                                errors.append("\"").append(name).append("\" in ").append(methodToString(method, parameterTypes, parameters)).append("\n");
-                            }
-                        }
+                int n = parameterTypes.length - 1;
+                // Ignore all the @Cached methods from our consideration.
+                while (n >= 0 && parameters[n].isAnnotationPresent(Cached.class)) {
+                    n--;
+                }
+                if (methodAnnotation.needsBlock()) {
+                    if (n < 0) {
+                        SUCCESS = false;
+                        Log.LOGGER.warning("invalid block method parameter position for " + method);
+                        continue;
                     }
+                    unguardedParams = unguardedParams | isParameterUnguarded(method, parameters, parameterTypes, n, errors);
+                    n--; // Ignore block argument.
                 }
-
-                if (unguardedObjectArgument) {
+                if (methodAnnotation.rest()) {
+                    if (n < 0 || !parameterTypes[n].isArray()) {
+                        SUCCESS = false;
+                        Log.LOGGER.warning("invalid rest method parameter " + n + " for " + method);
+                        continue;
+                    }
+                    n--; // ignore final Object[] argument
+                }
+                for (int i = 0; i < opt; i++, n--) {
+                    if (n < 0) {
+                        SUCCESS = false;
+                        Log.LOGGER.warning("invalid optional paramenter count for " + method);
+                        continue;
+                    }
+                    unguardedParams = unguardedParams | isParameterUnguarded(method, parameters, parameterTypes, n, errors);
+                }
+                // required arguments are not checked as they should not receive a NotProvided instance.
+                if (unguardedParams) {
                     SUCCESS = false;
                     Log.LOGGER.warning("ambiguous optional argument in " + node.getCanonicalName() + ":\n" + errors);
                 }
             }
         }
     }
 
+    private static boolean isParameterUnguarded(Method method, Parameter[] parameters, Class<?>[] types, int n, StringBuilder errors) {
+        Class<?> parameterType = types[n];
+
+        Parameter parameter = parameters[n];
+        boolean isNamePresent = parameter.isNamePresent();
+        if (!isNamePresent) {
+            AVAILABLE = SUCCESS = false;
+            Log.LOGGER.warning("method parameters names are not available for " + method);
+            return false;
+        }
+        String name = parameter.getName();
+
+        // The logic here is that an optional that are not of type Object are automatically guarded in the
+        // generated code, so we only need to check arguments of type Object.
```

Comment:
That's still unclear to me. Mentioning NotProvided seems important since that's what this whole checker is about.
`generated code` confuses me, let's talk about the type relations instead.